transport/transport.go

```diff
@@ -0,0 +1,95 @@
+package transport
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/mux"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// DialTimeout is the maximum duration a Dial is allowed to take.
+// This includes the time between dialing the raw network connection,
+// protocol selection as well the handshake, if applicable.
+var DialTimeout = 60 * time.Second
+
+// AcceptTimeout is the maximum duration an Accept is allowed to take.
+// This includes the time between accepting the raw network connection,
+// protocol selection as well as the handshake, if applicable.
+var AcceptTimeout = 60 * time.Second
+
+// UpgradedConn is a connection that is is an extension of the net.Conn interface that provides multiaddr
+// information, and an accessor for the transport used to create the conn
+type UpgradedConn interface {
+	mux.MuxedConn
+	network.ConnSecurity
+	network.ConnMultiaddrs
+
+	// Transport returns the transport to which this connection belongs.
+	Transport() Transport
+}
+
+// Transport represents any device by which you can connect to and accept
+// connections from other peers. The built-in transports provided are TCP and UTP
+// but many more can be implemented, sctp, audio signals, sneakernet, UDT, a
+// network of drones carrying usb flash drives, and so on.
+type Transport interface {
+	// Dial dials a remote peer. It should try to reuse local listener
+	// addresses if possible but it may choose not to.
+	Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (UpgradedConn, error)
+
+	// CanDial returns true if this transport knows how to dial the given
+	// multiaddr.
+	//
+	// Returning true does not guarantee that dialing this multiaddr will
+	// succeed. This function should *only* be used to preemptively filter
+	// out addresses that we can't dial.
+	CanDial(addr ma.Multiaddr) bool
+
+	// Listen listens on the passed multiaddr.
+	Listen(laddr ma.Multiaddr) (Listener, error)
+
+	// Protocol returns the set of protocols handled by this transport.
+	//
+	// See the Network interface for an explanation of how this is used.
+	Protocols() []int
+
+	// Proxy returns true if this is a proxy transport.
+	//
+	// See the Network interface for an explanation of how this is used.
+	// TODO: Make this a part of the go-multiaddr protocol instead?
+	Proxy() bool
+}
+
+// Listener is an interface closely resembling the net.Listener interface.  The
+// only real difference is that Accept() returns Conn's of the type in this
+// package, and also exposes a Multiaddr method as opposed to a regular Addr
+// method
+type Listener interface {
+	Accept() (UpgradedConn, error)
+	Close() error
+	Addr() net.Addr
+	Multiaddr() ma.Multiaddr
+}
+
+// Network is an inet.Network with methods for managing transports.
+type TransportBasedNetwork interface {
```

Comment:
```suggestion
type TransportNetwork interface {
```
Which ever name is chosen, the comment needs updating as well

Reply:
`TransportBasedNetwork` _is_ a mouthful, yet `TransportNetwork` lacks definition. But I take it.